File: src/app/api/admin/clear-manifests.ts
import { Router } from '@oak/oak';
import { clearAndEnsureDirectoryExists } from '../../../lib/file-system/clear-and-ensure-dir.ts';
import { returnToSender } from '../../../lib/http/return-to-sender.ts';
import { baseMiddlewares, protectedMiddlewares } from '../../base-middlewares.ts';
import { getShareManifestsDir } from '../../files/cache-folder.ts';
import { logger } from '../../logging/loggers.ts';
import { ensurePermissions } from '../../security/api-protect.ts';
import { csrfProtect } from '../../security/csrf-protect.ts';

export function register(router: Router) {
    router.post(
        '/api/admin/clear-manifests',
        baseMiddlewares(),
        ...protectedMiddlewares(),
        csrfProtect,
        ensurePermissions((p) => p.allowHousekeeping),
        async (ctx) => {
            await clearAndEnsureDirectoryExists(getShareManifestsDir());
            logger.info('Cleared share link manifests directory');
            returnToSender(ctx, {
                search: {
                    'message': 'Share link manifests cleared',
                },
            });
        },
    );
}

File: src/goapp/api/admin/clear_manifests.go
package admin

import (
	approutes "catears/ffs/goapp/app-routes"
	"catears/ffs/goapp/appmiddlewares"
	"catears/ffs/goapp/cache"
	"catears/ffs/goapp/resources"
	"catears/ffs/lib/router"
	"log"
	"net/http"
)

type clearManifestRouter struct{}

func (self *clearManifestRouter) Register(appRouter router.Router) {
	r := appRouter.With(
		appmiddlewares.CsrfProtect,
		appmiddlewares.EnsureClaim(resources.AdminResource, resources.HousekeepingClaim))

	r.Post("/api/admin/clear-manifests", self)
}

func (*clearManifestRouter) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	err := cache.ShareManifestsFolder.Recreate()
	if err != nil {
		w.WriteHeader(http.StatusInternalServerError)
		return
	}

	log.Println("Cleared share link manifests directory")
	router.ReturnToSender(w, r, router.Param("message", "Share link manifests cleared"))
}

func init() {
	approutes.Routes = append(approutes.Routes, &clearManifestRouter{})
}
